Feature/automatic parsing

tomereli opened the pull request:
This PR completes the automatic TLV parsing feature by enabling `cmdu_rx.parse()` to *always* parse the entire CMDU. As a result, the automatic parser `CmduTlvParser` is blindely parsing the entire CMDU, including WSC internal Attribute Lists for M1 and M2 messages.
Vendor specific inner classes are *not* automatically parsed since they may differ between different vendors.
 
Fixes #609 
Fixes #318

Commit:
common: set automatic parsing (breaks operational)

This commit updates the SocketThread and TransportSocketThread base
classes to always do a full parse of all the TLVs in the CMDU by calling
the CmduTlvParser.

This builds but breaks operational mode since all callers need to be
changed not to do a second addClass based on the message type.

We do not squash the fixes to all callers to make this reviewable,
therefore followups commits will "fix" the broken parts till we get back
to fully operational.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
automatic parsing - fix autoconfig search

Now that we automatically parse in the base class, replace all the
cmdu_rx.addClass calls for autoconfig search to cmdu_rx.getClass.
As a good side-effect, we no longer need a while loop to make sure we
are parsing the TLVs in order.

This fixes the autoconfiguration search flow between the controller and
agent.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: CmduTlvParser do not parse WSC TLV

The WSC TLV will be parsed by the caller, the same way it is done for
vendor specific TLVs.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
autoparse: fix WSC autoconfig

Now that we always parse in the base class, update autoconfiguration in
the controller and agent to use getClass<> instead of addClass<>.

While we're at it, we can finally get rid of the while loop used for
looping through the TLVs in the autoconfiguration handling by checking
which type of WSC message it is in the agent and controller.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: CmduMessageRx remove CmduHeader double swap

Preparative commit.

In CmduMessage::getMessageType() method, we are using the is_finalized()
getter to check if swap is needed for the header. The reason for this is
that sometimes we need to get the message type (or message id) after the
CMDU header was finalized, and sometimes we need it before it is
finalized. For this reason, the CmduMessage getter APIs are using the
is_finalized() method to check if swap is needed.

In the CmduMessageRx case, addClass<cCmduHeader>() will do swapping, but
since we always consider RX CMDUs as finalized (see BaseClass
constructor) CmduMessage::getMessageType() and
CmduMessage::getMessageId() will always swap the value back.
This is the reason that we have the extra swap to the cmdu header in
CmduMessageRx::parse().

This is obviousely not needed so this commit removes the extra header
swap and updates the m_finalized__ to be false by default when m_parse__
is set.

However, when forwarding CMDUs, the second parse will now fail since the
header is already swapped, so the second header parse will swap back to
network byte order when we expect it to be in host byte order.
The previous flow worked by sheer luck (!) since we always swapped the
header on rx no matter what. So fix this as well by adding the needed
cmdu_rx.swap() in the backhaul manager, which is the only component
which still forwards CMDUs (from son slave to controller and back).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Enable auto-parsing by default

Update CmduMessageRx::parse() to always parse all TLVs.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: ClassList swap also inner class list

Preparative commit.

When ClassList::swap() is called explicitely, it might need to also swap
the inner class list, which till now it didn't. This is required only if
the end point which actually parsed the data also called the inner class
list parse which in turn swapped the TLV payload.

This commit may be redundant for now but it doesn't hurt to swap the
inner class list if such exists - since it means that the payload data
needs to be swapped.

The actual cases where cmdu_rx.swap() is called explicitly is in the
backhaul manager when forwarding CMDUs from the son slaves to the
controller and vice versa, and in print_cmdu_types().

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: delete unrelated comment

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide - replace addClass with getClass for rx

Mechanical change (done manually) - replace all cmdu_rx.addClass<T>()
with cmdu_rx.getClass<T>() since now parse already happened and calling
addClass again is a bug.

For the cases where multiple classes of the same type exists, use range
based loop:

for (auto i : cmdu_rx.msg.getClassList<T>())
	// do something with i

As part of this change, code sections which used the while loop in order
to call addClass<T> in the right order are no longer needed and simply
removed. For example, see handle_channel_selection_request().

With this change, prplMesh passes all test_flows.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add class_cast for parsing unknown tlvs

Preparative commit.
Now that parse() always parses the entire message, we cannot support any
non-standard TLVs such as the ones defined specifically for testing such
as tlvVarList used all over the tlvf_test.
This is because the default parser (which in followup commits will
become the only parser) is unaware of these TLV types, and therefore
treats them as Unknown TLVs.

In order to support non standard TLVs (and tlvf classes) on parsing, we
need to be able to do one of the following:
1. introduce these TLV types to the parser
2. add non 1905 parsers infractructure
3. add support for class_cast

Option 1 is out of the question since we don't want to pollute the 1905
namespace with non standard TLVs which are only used for testing at the
moment.

Option 2 is probably the best one but an overkill since we don't really
need this kind of infrastructure at this point.

Option 3 is the simplest - intruduce the class_cast which was proposed
in #481 and rejected since there was no need for it then, but now it can
be used to cast the unknown TLVs which the default parser parses.

Therefore, this commit brings back option 3 - class_cast - and adds it
to BaseClass. It will be used in a followup commit which fixes the
tlvf_test after the parse changes.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: test: fix test_complex_list()

This commit uses the class_cast<T>() method to cast the tlvUnknown added
by the automatic parser.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: test: fix test_all()

- Remove all getNextTlvType(type) checks which are irrelevant now since
the checks are done as part of received_message.parse()
- Remove header protection test since it is covered by the parse
function now.
- Replace all received_message.addClass<T> with
received_message.getClass<T>
- Remove received_message.addClass<tlvUnknown>() (introduced in commit
a43e65e0d1913873655d140098932eef7149963e) since the unknown class
is tlvMacAddress which is known to the default parser. No need to
replace it since the next TLV is also unknown - TlvTestVarList which is
unknown to the parser.
- Remove double parse for M2 - replace with
std::dynamic_pointer_cast<WSC::m2>(WSC::AttrList::parse(*tlv))

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -311,116 +311,74 @@ bool master_thread::handle_cmdu_1905_1_message(const std::string &src_mac,
-        LOG(ERROR) << "tlvAlMacAddressType missing - ignoring autconfig search message";
+    auto tlvAlMacAddressType = cmdu_rx.getClass<ieee1905_1::tlvAlMacAddressType>();
+    if (!tlvAlMacAddressType) {
+        LOG(ERROR) << "getClass<tlvAlMacAddressType> failed";
```
Really nitpicking here, but one can argue that this is not really an error.
I believe that errors should be reported when something went so wrong that regular operation can't be continued.
In this case, it's probably a malformed CMDU or a missing TLV, which doesn't really impact the Controller's operation, so a WARNING would be more appropriate here.

tomereli replied:
I disagree - I think that errors are anything not expected which changes the way the system behaves. In this case, an agent tried to onboard our controller and we simply didn't answer it due to some missing TLV - we should easilly find this in the logs while looking for either error or warning, so changing to warning is OK, but not as part of this PR since it was ERROR before this change.

IMHO FATAL is more appropriate than error to describe an unrecoverable error, not ERROR.

vitalybu commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -311,116 +311,74 @@ bool master_thread::handle_cmdu_1905_1_message(const std::string &src_mac,
+
+    LOG(DEBUG) << "searched_role=" << int(tlvSearchedRole->value());
+    if (tlvSearchedRole->value() != ieee1905_1::tlvSearchedRole::REGISTRAR) {
+        LOG(ERROR) << "invalid tlvSearchedRole value";
```
Continuing my previous comment, this is probably a warning as well.
Also, a good idea would be to print the received value... Will help debugging.

tomereli replied:
Same anwer - perhaps you're right, but I didn't change the error logging as part of this PR and I don't want to do that either.

vitalybu commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3848,77 +3848,47 @@ bool slave_thread::handle_autoconfiguration_renew(Socket *sd, ieee1905_1::CmduMe
+    // it will be published back on the local bus because the destination is our
+    // AL-MAC, and the agent does listen to this CMDU message type).
+    // If this is a looped back M1 CMDU, we can treat is as handled successfully.
+    if (cmdu_rx.getClass<wfa_map::tlvApRadioBasicCapabilities>())
```
Shouldn't we check that it's indeed our looped request?
What about non-local agents (repeaters) that send an autoconfig message to our controller?
The local agent won't receive the messages?

tomereli replied:
It doesn't matter if its our looped request or an M1 CMDU from another agent, in both cases we are dropping the message.

tomereli replied:
We do get the other agents WSC messages and the handling is and should be the same as looped back M1 CMDUs - we should ignore them. Checking that this is indeed our looped request is a much harder thing to do with ZMQ which is our current MSGLIB backend. Perhaps it will become easier once we switch to a different implementation, but for now, this is the only way we can do that.

Galvien commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3862,77 +3862,47 @@ bool slave_thread::handle_autoconfiguration_renew(Socket *sd, ieee1905_1::CmduMe
     if (m2_list.empty()) {
         LOG(ERROR) << "No M2s present";
         return false;
     }
```
In my opinion, instead of iterating over the cmdu_rx with `getClassCount<T>()`  we should implement a function named something like `getAllClasses<T>()` (`getClassList<T>()` would be too confusing) that returns a vector of all tlvs of type T in the cmdu.
- Firstly, it would be much more befitting of the OOP oriented direction of this pull request and it's predecessors.
- Secondly, once implemented, we could iterate over the vector with a foreach loop, making boilerplates such as this one shorter and more readable.

tomereli replied:
Good idea - done.

vitalybu commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -19,18 +19,13 @@ CmduMessageRx::CmduMessageRx(uint8_t *buff, size_t buff_len) : CmduMessage(buff,
     if (!cmduhdr)
         return false;
 
-    cmduhdr->class_swap(); //swap back cmduheader
```
Was is swapped twice? How did it work before?

tomereli replied:
Updated commit message with full explanation.

arnout replied:
The improved commit message is excellent!

Galvien reviewed (commented):
1. This PR is a continuation of a few previous and more importantly - _recently merged_ PRs, I think it would be nice if it was mentioned in the PR comment, and even have references to relevant commits (especially cool if it's a github comment, as github provides the direct link to the commit page) when relevant.
2. When performing a tree-wide change, such as in febf29c, please provide references to the changes worth taking a look at. The changes replacing all of the `addClass<T>()` calls with `getClass<T>()` are pretty ignorable, but there are some more important, logic changes that should be noted.
Some suggested methods of highlighting such cases:
    - Split the commit into a 'boilerplate changing' commit and a 'meaty' commit.
    - In the commit message, add a suffix along the lines of "note line #1234"
    - On github, add comments on code parts worth checking out, and mention it in the commit message.

Galvien commented on framework/tlvf/src/src/CmduTlvParser.cpp:
```diff
@@ -108,7 +108,7 @@ std::shared_ptr<BaseClass> CmduTlvParser::parseNextTlv()
-            WSC::AttrList::parse(*tlv);
+            tlv->addInnerClassList(WSC::AttrList::parse(*tlv));
         return tlv;
     }
```
CmduTlvParser shouldn't handle special cases such as this one, the attribute list parsing should be done as part of `tlvWsc::addClass()`, this requires a TLVF change, as tlvWsc is an autogenerated class.
It's still worth mentioning in my opinion, and I will create an issue for that.

tomereli replied:
I generally agree, TLVF should add meta option to describe which inner class list should be part of each TLV and then we can autogenerate the parser to call the `addInnerClassList` but Its an overkill IMO at this stage.

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [TLVF Automatic parsing full support](https://trello.com/c/y3wISsqB/22-tlvf-automatic-parsing-full-support)

Commit:
tlvf: ClassList cleanups - delete unused methods

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Move CmduTlvParser into CmduMessageRx class

Now that the automatic parsing is complete and we know that parsing is
always done using the default and only parser which is the 1905_1
parser, we can remove the useless infrastructure that was added in
commit e0e6a1d and move the parser to be a private method in the
CmduMessageRx class called by the CmduMessage::parse() method.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add getClassList<T>()

Preparative commit.
Add template function getClassList<T>() which returns
std::list<std::shared_ptr<T>> of all the classes of type T in the
m_class_vector.
This will be used for range based loops for iterating through all the
classes of type T as follows:

for (auto i : msg.getClassList<T>)
	// do something with i

This will be used in all cases where more than one class of the same
type exists in the vector - and will be used instead of addClass done by
callers when parse will parse the entire message in the following
commits.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on framework/tlvf/src/include/tlvf/ClassList.h:
```diff
@@ -105,6 +106,23 @@ class ClassList {
+     * @tparam T class template
+     * @return std::list<std::shared_ptr<T>> list of class T pointers
+     */
+    template <class T> std::list<std::shared_ptr<T>> getClassList() const
```
I know that @vitalybu suggested to use a std::list instead of a std::vector because it's more efficient, but he's wrong according to [CppCoreGuildelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon2-prefer-using-stl-vector-by-default-unless-you-have-a-reason-to-use-a-different-container). std::vector has

- the fastest general-purpose access (random access, including being vectorization-friendly);
- the fastest default access pattern (begin-to-end or end-to-begin is prefetcher-friendly);
- the lowest space overhead (contiguous layout has zero per-element overhead, which is cache-friendly).

Specifically in this case, where we expect the result set to contain 0, 1 or 2 elements and very rarely more than that, a vector is more efficient on all counts.

Obviously, though, since the result is usually small, the impact of this choice is not significant. So nitpick.

tomereli replied:
I think that we can stick with class list since it is in fact a class list, and since the impact is not significant.

arnout commented on common/beerocks/tlvf/src/src/beerocks_message.cpp:
```diff
@@ -61,7 +61,7 @@ message_com::parse_intel_vs_message(ieee1905_1::CmduMessageRx &cmdu_rx)
         LOG(ERROR) << "cmdu is not initialized!";
     }
-    auto tlv = cmdu_rx.addClass<ieee1905_1::tlvVendorSpecific>();
+    auto tlv = cmdu_rx.getClass<ieee1905_1::tlvVendorSpecific>();
```
It is weird that the commit message says that addClass will be converted to getClass in later commits, but then this one change is done in the same commit.

Also, why is the parse call on line 87 not converted in the same commit?

tomereli replied:
You're right, this is the first time we change addClass to getClass so that the initial message passing of vendor speficic CMDUs will work. I will add this in a separate commit which will come right after this one.
Regarding line 87 - will fix this also.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -311,116 +311,74 @@ bool master_thread::handle_cmdu_1905_1_message(const std::string &src_mac,
-    } else {
-        LOG(ERROR) << "tlvAlMacAddressType missing - ignoring autconfig search message";
+    auto tlvAlMacAddressType = cmdu_rx.getClass<ieee1905_1::tlvAlMacAddressType>();
+    if (!tlvAlMacAddressType) {
```
The removal of the loop and moving the checks immediately after the declaration is a bit more than what is strictly needed for "automatic parsing - fix autoconfig search".

In fact, I would have split the commits completely differently: first commit that just changes the default value of the parse_tlvs parameter from false to true, and in the same commit replace all cmdu_rx.addClass with cmdu_rx.getClass. That's an almost purely mechanical change which probably doesn't break anything. Then subsequent commits can de improvements and cleanups, like this one.

Anyway, very much nitpick.

**EDIT**

D'oh, I was stupid. Just replacing addClass with getClass *doesn't* fix things, because the while loop no longer works correctly (after parse, getNextTlvType will always return EOM). So never mind.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3862,77 +3862,44 @@ bool slave_thread::handle_autoconfiguration_renew(Socket *sd, ieee1905_1::CmduMe
+        auto m2 = std::dynamic_pointer_cast<WSC::m2>(tlv->getInnerClassList());
+        if (m2)
+            m2_list.push_back(*m2);
+    }
```
I don't like this very much. I would much prefer if we would do the recursive parsing directly here instead of in `CmduTlvParser::parse`. So:

```
std::vector<WSC::m2> m2_list;
for (auto tlv : cmd_rx.msg.getClassList<ieee1905_1::tlvWsc>()) {
    auto m2 = WSC::m2::parse(tlv);
    if (!m2) {
        LOG(INFO) << "Not a valid M2 - ignoring WSC CMDU";
        return false;
    }
    m2_list.push_back(m2);
}
```

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3862,77 +3862,44 @@ bool slave_thread::handle_autoconfiguration_renew(Socket *sd, ieee1905_1::CmduMe
+    for (auto tlv : cmdu_rx.msg.getClassList<ieee1905_1::tlvWsc>()) {
+        auto m2 = std::dynamic_pointer_cast<WSC::m2>(tlv->getInnerClassList());
+        if (m2)
+            m2_list.push_back(*m2);
```
I think we need to return false if m2 is not valid, like we did before.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3988,9 +3955,10 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
+    if (beerocks_header) {
         LOG(INFO) << "Intel controller join response";
-        if (!parse_intel_join_response(sd, cmdu_rx)) {
+        if (!parse_intel_join_response(sd, *beerocks_header)) {
```
It would have been nice to have done this change in a separate, preparative commit (removing intel_agent and relying on `if (beerocks_header)` is something that can be done independent of the change to `parse(true)`).

nitpick, of course.

tomereli replied:
Splitting now is too much extra effort but I agree, will take it into account next time.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3862,77 +3862,44 @@ bool slave_thread::handle_autoconfiguration_renew(Socket *sd, ieee1905_1::CmduMe
+    // which was just looped back (when we send the M1 to the local controller,
+    // it will be published back on the local bus because the destination is our
+    // AL-MAC, and the agent does listen to this CMDU message type).
+    // If this is a looped back M1 CMDU, we can treat is as handled successfully.
```
Actually, we could just drop this special handling entirely now. The return value is anyway ignored (it gets passed up to socket_thread::work without further comment, and there it essentially gets ignored). We really should be checking for M1/M2, which we do implicitly by checking m2_list.empty().

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -703,60 +703,28 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(const std::string &sr
+    // AL-MAC, and the controller does listen to this CMDU message type).
+    // If this is a looped back M2 CMDU, we can treat is as handled successfully.
+    if (cmdu_rx.getClass<wfa_map::tlvApRadioIdentifier>())
+        return true;
```
Same here, I think we don't need this special handling and we can just rely on the m1 check.

tomereli replied:
Correct, removed.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -703,60 +703,28 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(const std::string &sr
         return false;
     }
-    auto m1 = std::dynamic_pointer_cast<WSC::m1>(WSC::AttrList::parse(*tlvWsc));
+    auto m1 = std::dynamic_pointer_cast<WSC::m1>(tlvWsc->getInnerClassList());
```
Again, it would make more sense to me do do the inner parsing here instead of doing a down-cast.

tomereli replied:
Done.

arnout reviewed (commented):
Reviewed up to "autoparse: fix WSC autoconfig". Submitting the review already so you can start processing the feedback. Up to now, I would approve it (all changes are either nitpicks or possibly to be fixed later).

arnout commented on framework/tlvf/src/src/ClassList.cpp:
```diff
@@ -34,6 +34,8 @@ void ClassList::swap()
     // call all tlv finalize functions
     for (auto const &c : m_class_vector) {
         c->class_swap();
+        if (c->getInnerClassList())
+            c->getInnerClassList()->swap();
```
There is also an explicit swap at the end of `message_com::print_cmdu_types`. I'm not even sure if it is appropriate there, but it should probably be mentioned in the commit message.

tomereli replied:
Actually - now that parse() does not automatically parse the inner class list, it isn't swapped, so this commit can be removed altogether - the backhaul manager parse OR the notorious print_cmdu_types method will simply not parse the inner list. I'm debating weather we should keep this here for the future in case we do the parsing then decide to swap back for forwarding. What do you think?

In any case I updated the commit message for now cause it was wrong.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1749,7 +1749,7 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
-    auto tlvHigherLayerData = cmdu_rx.addClass<wfa_map::tlvHigherLayerData>();
+    auto tlvHigherLayerData = cmdu_rx.getClass<wfa_map::tlvHigherLayerData>();
     if (!tlvHigherLayerData) {
         LOG(ERROR) << "addClass wfa_map::tlvHigherLayerData failed";
```
There are still quite a lot of log messages that mention addClass instead of getClass. nitpick.

tomereli replied:
Arr yep.. We can fix that in a mechanical change I guess. Anyway, different PR.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1553,51 +1474,22 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+    if (!client_association_event_tlv) {
+        LOG(ERROR) << "addClass wfa_map::tlvClientAssociationEvent failed";
+        return false;
+    }
```
This is not correct. The original code had the following, which can just stay:

```
   if (!client_association_event_tlv) {
        // The standard allows having zero TLV_CLIENT_ASSOCIATION_EVENT, in that case, there is no
        // point to continue the handling of this message.
        LOG(INFO) << "wfa_map::tlvClientAssociationEvent not found";
        return true;
    }
```

arnout commented on framework/tlvf/src/include/tlvf/BaseClass.h:
```diff
@@ -36,6 +36,14 @@ class BaseClass {
     bool is_finalized() { return m_finalized__; };
     void addInnerClassList(std::shared_ptr<ClassList> list) { m_inner__ = list; };
     std::shared_ptr<ClassList> getInnerClassList() { return m_inner__; };
+    template <class T> std::shared_ptr<T> class_cast()
```
Please add documentation for new functions. In this case, it should be in framework/tlvf/README.md because that's where the rest of the documentation is (though personally I'd prefer to have everything as doxygen).

tomereli replied:
So lets start with updating the header only, we can remove the documentation later on from the README.md since it is obsolete. @CoralMalachi can you do that as part of the update to the README.md file?

arnout commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -7,28 +7,176 @@
 using namespace ieee1905_1;
 
-CmduMessageRx::CmduMessageRx(uint8_t *buff, size_t buff_len) : CmduMessage(buff, buff_len)
+std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
```
This change goes a bit against the automatically generated parser that @Galvien still has lined up AFAIU. A step in the right direction would be to make this an external function that takes the ClassList as an argument and that open-codes getNextTlvType, because that's what the generated code would look like.

Nitpick.

tomereli replied:
I don't think we need an automatically generated parser anymore. We can add the missing TLVs for R2 manually, it shouldn't take this long - how many are there?

arnout replied:
Hehe. There are about 30 new TLVs in R2... I agree though that the effort of maintaining that manually is pretty small. I just thought, since @Galvien has the code anyway, we could easily merge it.

tomereli replied:
The code needs to be reviewed again and added to TLVF so it will take Gal at least half a day I assume, and I rather he work on KW issues for now. We can use his branch to regenerate the R2 TLVs when the time comes, but if someone has the time and will to do that, I won't object to merging the automatic parser, its just its really nice to have and not a must.

Galvien replied:
Right now the code serves more as a method of quickly generating parsing code later to be copied and integrated manually into the codebase rather than a mechanism for creating autogenerated parser files.
So the course of action regarding adding new automatic parsers would be running TLVF once with the required configuration for the autogenerator, then copying it's output manually, the whole process should take 5-10 minutes instead of the hours needed for integrating it fully (and the far more difficult aspect - properly) into TLVF, that's at least where it stands ATM...

arnout reviewed (approved):
One important change that really needs to be fixed before the merge: client_association_event_tlv

Also, the commit message of 'tlvf: test: fix test_all() ' has a bunch of spelling mistakes, it would be nice to correct those.

Since all changes can be handled without re-review, I'm already approving, but mark as donotmerge.

Commit:
autoparse: fix parse_intel_vs_message

Replace addClass with getClass for getting the vendor specific TLV.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
